File: 0_string/2022_1_QE_1.py
'''
- implement max_palindrome(s) 
- returns a list of substrings of s that are maximal palindromes 
- the list contains palidromes that are not a substring of another palindrome
- ex) 
For s = "kabccbadzdefgfeda",
max palindromes(s) should return ["k", "abccba", "dzd", "defgfed"]

For s = "kabccba dzabccbaza",
max palindromes(s) should return ["k", " ", "d", "zabccbaz", "aza"]
'''

# problem (a) 
# a function checks if the string s is a palindrome 
def palindrome(s:str) -> bool:
    if len(s) == 1: 
        return True
    len_s = len(s) 
    mid = len_s // 2 
    if len_s % 2 == 0:
        l = mid -1
        r = mid 
    else:
        l = r = mid 

    while l >= 0 and r < len_s and s[l] == s[r]:
            l -= 1 
            r += 1 

    if l < 0 :
         return True 
    return False 



# problem (b) 
# a function substring(s, t) that checks if the string t is a substring(s,t) 

def substring(s:str,t:str) -> bool:
    if len(s) >= len(t):
        len_t = len(t) 
        difference_len = len(s) - len(t) 
        for idx, char in enumerate(s[:difference_len]):
            if s[idx:idx+len_t] == t:
                return True 

        return False 
    else:
        return False 




# problem (c) 
# funtion max_palindromes(s) that uses palidrome(s) and substring(s,t) 
def max_palindrome(s:str) -> list:
    temp = []

    # find all palindromes 
    for i in range(len(s)):
        for j in range(i, len(s)+1):
            candidate_s = s[i:j]
            if palindrome(candidate_s):
                temp.append(candidate_s) 
    
    for _ in temp:
        print(temp)

    # find whether there are some substring or the same palindrome
    result = set(temp) 
    result = list(result)
    result.remove('')
    # print(result, len(result))

    # combination not permutation since you do the opposite in the 2 consecutive if prahse below
    # remove_list = []

    # for i in range(len(result)-1):
    #     for j in range(i+1, len(result)):
    #         # print(i,j)
    #         if substring(result[i], result[j]) == True:
    #             # result.remove(result[j]) # result의 길이가 계속 변해서 안됨 
    #             remove_list.append(result[j])
    #         if substring(result[j], result[i]) == True:
    #             # result.remove(result[i])
    #             remove_list.append(result[i])
    # print(remove_list)
    # print(result)
    # #subtract 
    # for rem in remove_list:
    #     result.remove(rem)
    
    def remove_redundant_substring(string_list):
        result = []
        ### important to sort first #### 
        string_list.sort(key=len) # sort the strings by length to process short substrings first 
    
        for i, s in enumerate(string_list):
            is_redundant = False 
            for j in range(i+1, len(string_list)):
                if substring(string_list[j],s):
                    is_redundant = True 
                    break 
            if not is_redundant:
                result.append(s) 
        return result 
    
    ans = remove_redundant_substring(result)

    return ans


            
if __name__ == '__main__':
    # s = "kabccbadzdefgfeda"
    s = "kabccba dzabccbaza"
    # t = 'sdvxz' 
    # print(substring(s,t))
    print(max_palindrome(s))
         

              

File: 0_string/2022_1_QE_1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_set>

using namespace std;

bool is_palindrome(const string& s) {
    int len_s = s.size();
    int mid = len_s / 2;
    int l;
    int r;

    if (len_s % 2 == 0) {
        l = mid - 1;
        r = mid;
    } else {
        l = r = mid;
    }

    while (l >= 0 && r < len_s && s[l] == s[r]) {
        l -= 1;
        r += 1;
    }
    
    return (l == -1) ? true : false;
}

// int main(){
//     string s = "my";
//     string k = "mystring";
//     string d = "d";
//     string e = " ";
//     string a = "qwerreq";
//     cout << is_palindrome(s) << " ";
//     cout << is_palindrome(k) << " ";
//     cout << is_palindrome(d) << " ";
//     cout << is_palindrome(e) << " ";
//     cout << is_palindrome(a) << " ";
    
//     return 0;
// }

bool substring(const string& s, const string& t) {
    int len_s = s.size();
    int len_t = t.size();

    if (len_s < len_t) {
        return false;
    }

    for (int i = 0; i <= len_s - len_t; ++i) {
        if (s.substr(i, len_t) == t) {
            return true;
        }
    }

    return false;
}

vector<string> max_palindrome(const string& s) {
    // find all palindromes 
    // vector<string> temp;
    unordered_set<string> temp; // insert, begin, end, empty, size, erase(iterator), extract(element)
    int len_s = s.size();
    for (int i = 0; i < len_s; i++){
        // for (int j = i+1; j <= s.size() +1; j++){
        for (int j = 0; j<=s.size()+1; j++){ // j는 i+1이 아니라 0부터여야함. 왜냐면 substring의 size를 나타내기 때문 
            string candidate = s.substr(i,j-1);
            // cout << '"' << candidate << '"' << " ";
            if (is_palindrome(candidate)){
                
                temp.insert(candidate);
            }
        }
    }
    
    //make set into vector 
    vector<string> target;
    for (const auto&it: temp){
        target.push_back(it);
    }
    
    // for (string _:target){
    //     cout << _ << " ";
    // }
    // cout << '\n';
    
    // remove redundant substring 
    sort(target.begin(), target.end(), [](const string &s, const string& t){
        return s.size() < t.size();
    });
    
    vector<string> ans;
    
    for (int i = 0; i <target.size(); i++){
        bool is_redundant = false;
        for (int j = i+1; j <target.size()+1; j++){
            if (substring(target[j], target[i])){
                is_redundant = true;
                break;
            }
        }
        if (!is_redundant){
            ans.push_back(target[i]);
        }
    }
    
    return ans;
}
int main() {
    string s = "kabccba dzabccbaza";
    
    vector<string> result = max_palindrome(s);

    for (const string& palindrome : result) {
        cout <<'"' << palindrome << '"' << " "; // ["k", " ", "d", "zabccbaz", "aza"]
    }

    return 0;
}

